### PR TITLE
4789 item variant UI (PT 1)

### LINKED ISSUE #4789: GAPS: Implement item variants: Item UI changes
Part of epic #4362

## Is your feature request related to a problem? Please describe 👀
There are several aspects of the GAPS work which require item variant capabilities
For reference this is the full [item variant PRD](https://docs.google.com/document/d/1fRZslRohRGaZnUo13a2eKZjfO0DGWwXtIFsbEah7A4Y/edit#heading=h.hyld495fuxe5)

We won't need all aspects of that; but we will need the ability to create and edit and sync item variants.

The variants will need to have these properties for the GAPS work:

| Name | Type   | Reference |
|:------ |:------ |:---- |
| Volume (x3) | float | #3917 |
| Bundle items | reference to n item variants | #3923 |
| Temperature range | enum | #4368 |

## Describe the solution you'd like 🎁

See #5141

UI - on OMS central:

image

### Handling stock

Done in separate issue: #5145

### PR DESCRIPTION
**Pull request by lache-melvin**
Fixes #4789

# 👩🏻‍💻 What does this PR do?

This is part one, it was getting big enough! Adds base UI for item variants, with a mocked API sending some dummy data.

Screenshot 2024-10-24 at 8 43 21 AM

Still need to do the save mutation/manage edit state, and the manufacturer search component.

Things that will be done in separate issues:
- cold storage type dropdowns #4368
- The packaging variants edit modal #3917
- Bundling #3923

## 💌 Any notes for the reviewer?

# 🧪 Testing

- [ ] On OMS central
- [ ] Items -> item -> there is a variants tab
- [ ] Add variant button works
- [ ] It looks kind of like the design
- [ ] When running not as central - no variants tab

# 📃 Documentation

- [x] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

File: client/packages/common/src/intl/locales/pt/common.json [hunk 1]
```diff
@@ -1032,5 +1032,72 @@
   "title.stock-on-hand": "Estoque Disponível",
   "tools": "Ferramentas",
   "warning.caps-lock": "Advertência: Caps Lock está ativada",
-  "warning.nothing-to-supply": "Nada para fornecer!"
+  "warning.nothing-to-supply": "Nada para fornecer!",
```
Comment (lache-melvin) [inline client/packages/common/src/intl/locales/pt/common.json:1035]: Ah sorry, ignore these, I think I pulled a more recent version of develop into my branch than the base of this PR 👀

**Comment by adamdewey** [comment]
Nice!

@lache-melvin  Instead of `Cold storage type` can we use something a little less restrictive such as `Storage temperature` ?

**Comment by adamdewey** [comment]
Actually, I think just `Temperature` would be fine and more concise!

When we have the rest of the data fields in there can we also get @mariyamsupply to cast her eye over how best to refine the layout? 🙏

**Comment by lache-melvin** [comment]
Oh yeah, I missed doses per unit 😅 FYI am going off this design: https://github.com/msupply-foundation/open-msupply/issues/4789#issuecomment-2427772421

**Comment by adamdewey** [comment]
Thanks @lache-melvin - apologies, I also added a comment after your commit that maybe `Temperature` would be better as it's more concise?

For the layout, I see - hmm, it sort of looks a bit confusing to me with the rows of the 2 tables not lining up but being next to each other without very clear delineation? 

I think likely that eventually there will be a lot of information in each item variant - for example when it starts being used for stock management, having barcodes associated with them etc. 

Am conscious we have limited time / budget in this release - wondering if there are some simple suggestions that @mariyamsupply could give? Otherwise I fear we'll condemn it to the same depressing fate as the Item page 😿

**Comment by lache-melvin** [comment]
@adamdewey Updated the screenshot to latest changes, but don't want to spend too long on this (it's central server only UI)

This is using Mariya's designs as best I can for now with the component library we have... hopefully in future we can iterate on designs before we get into dev time! 😅

**Comment by adamdewey** [comment]
Ah it's central UI only?

Ok, will be interested to see how item variants are displayed to the standard user!

Totally agree it would be good to review designs before getting to dev stage, if only we had more time 😅

Thanks Laché!

File: client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemVariantsTab.tsx [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+} from '@openmsupply-client/common';
+import { ItemPackagingVariantsTable } from './ItemPackagingVariantsTable';
+import { ItemVariantFragment, PackagingVariantFragment } from '../../../api';
+
+const dummyPackVariants: PackagingVariantFragment[] = [
+  {
+    __typename: 'PackagingVariantNode',
+    id: '1',
+    packagingLevel: 1,
+    name: 'Primary',
```
Comment (jmbrunskill) [inline client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemVariantsTab.tsx:28]: I was thinking we might have a mapping something like `label.packaging_level.1` = `Primary` so it was translatable, and potentially could still allow the user to still set a name, but I think this is good. We might need some way to default to this when we add new variants I guess?

File: server/graphql/types/src/types/item.rs [hunk 3]
```diff
@@ -133,6 +135,19 @@ impl ItemNode {
         Ok(result)
     }
 
+    // TODO: maybe do this in a loader!
```
Comment (jmbrunskill) [inline server/graphql/types/src/types/item.rs:138]: Loader does sound useful. I think we need to think about how we can split out our types though, as in some cases you won't want to load all the extra stuff in on item (if you just need name and code for example). I'm  not sure we have a good pattern established for that?

File: server/graphql/types/src/types/item_variant.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+#[Object]
+impl ItemVariantNode {
+    pub async fn id(&self) -> &String {
+        &self.item_variant.id
+    }
+
+    pub async fn item_id(&self) -> &String {
+        &self.item_variant.item_link_id // TODO join to item for item_id
```
Comment (jmbrunskill) [inline server/graphql/types/src/types/item_variant.rs:20]: Just do an item loader? Not sure what we'll actually use this for?

I think I remember @andreievg saying that repository or service layer should map the item_link_id to item_id.
Is there a wiki about this?

Not for this PR but I realise I'm not clear on how we are doing it :)

File: server/graphql/types/src/types/item_variant.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    pub async fn name(&self) -> &String {
+        &self.item_variant.name
+    }
+
+    pub async fn doses_per_unit(&self) -> &Option<f64> {
+        &self.item_variant.doses_per_unit
+    }
+
+    pub async fn manufacturer_id(&self) -> &Option<String> {
+        &self.item_variant.manufacturer_link_id // TODO join to name for manufacturer_id
```
Comment (jmbrunskill) [inline server/graphql/types/src/types/item_variant.rs:32]: Name loader?

**Review by jmbrunskill — APPROVED** [review]
Looks great to me, lets ship it!

File: client/packages/system/src/Item/DetailView/DetailView.tsx [hunk 6]
```diff
@@ -44,11 +47,16 @@ export const ItemDetailView: FC = () => {
...
-  Environment.FEATURE_PACK_VARIANTS &&
+  isCentralServer &&
     tabs.push({
-      Component: <PackVariantsTab itemId={data.id} />,
-      value: t('label.pack-variants'),
-    });
+      Component: <ItemVariantsTab itemVariants={data.variants} />,
+      value: t('label.variants'),
+    }),
+    Environment.FEATURE_PACK_VARIANTS &&
```
[the remark this answers is not in the document]
Comment (lache-melvin) [reply client/packages/system/src/Item/DetailView/DetailView.tsx:55]: Ah yeah definitely!

File: client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemVariantsTab.tsx (hunk 2 above)
Comment (lache-melvin) [reply client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemVariantsTab.tsx:28]: Yep, my thinking when implementing the `add` button properly was to have these translatable - hadn't thought about them being able to change the name though - I suppose as long as we display the packaging level then we could!

File: server/graphql/types/src/types/item.rs (hunk 3 above)
Comment (lache-melvin) [reply server/graphql/types/src/types/item.rs:138]: What do you mean? If we use the loader then these wouldn't get loaded if only selecting a subset of fields..

Unless you're saying we should add another query to get the variants directly. Which probably, though I think the current cases we have, we'd only be querying the variants in the context of an item 🤷‍♀️

File: server/graphql/types/src/types/item_variant.rs (hunk 4 above)
Comment (jmbrunskill) [reply server/graphql/types/src/types/item_variant.rs:20]: I think my point was if we're generally loading the item_variant from a loader on item, looking up or even providing item_id on the node feels a bit redundant. I guess, we just wait to see what we actually need for the front end. Maybe we can even just get rid of this for now until we need it.

Definitely keen for that Loader write up. `link_id` one would be good too!

File: server/graphql/types/src/types/item.rs (hunk 3 above)
Comment (jmbrunskill) [reply server/graphql/types/src/types/item.rs:138]: I take it back this is already being done.
I see in Item API we have 

```
fragment ItemRow on ItemNode {
  __typename
  id
  code
  name
  unitName
}

fragment ItemWithPackSize on ItemNode {
  ...ItemRow
  defaultPackSize
}
fragment ItemStockOnHand on ItemNode {
  ...ItemWithPackSize
  availableStockOnHand(storeId: $storeId)
}
fragment ItemRowWithStats on ItemNode {
  ...ItemStockOnHand
  stats(storeId: $storeId) {
    __typename
    averageMonthlyConsumption
    availableStockOnHand
    availableMonthsOfStockOnHand
    totalConsumption
  }
}
```

So I think we already have what I was thinking of.
Just didn't want the loader running unless we actually need it in the context...

**Pull request merged**